### PR TITLE
expression.d: Use foreach

### PR DESCRIPTION
**Pull request by wilzbach**
~~~While this is straight-forward,~~~ I have to admit that it can't be automatized (easily).

### CODE REVIEW

File: src/dmd/expression.d [hunk 1]
```diff
@@ -1769,9 +1764,8 @@ extern (C++) abstract class Expression : RootObject
         {
             a = new Expressions();
             a.setDim(exps.dim);
-            for (size_t i = 0; i < a.dim; i++)
+            foreach (i, e; *exps)
             {
-                Expression e = (*exps)[i];
                 (*a)[i] = e ? e.syntaxCopy() : null;
```
Comment (wilzbach) [inline src/dmd/expression.d]: Would be could to have `lockstep` / `zip` here, s.t. we can drop the iteration index in favor of the second range to be iterated.

File: src/dmd/expression.d [hunk 2]
```diff
@@ -3499,9 +3492,8 @@ extern (C++) final class TupleExp : Expression
                 return false;
             if (e0 && !e0.equals(te.e0) || !e0 && te.e0)
                 return false;
-            for (size_t i = 0; i < exps.dim; i++)
+            foreach (i, e1; *exps)
             {
-                Expression e1 = (*exps)[i];
                 Expression e2 = (*te.exps)[i];
```
Comment (wilzbach) [inline src/dmd/expression.d]: Another potential winner for `lockstep`

File: src/dmd/expression.d [hunk 3]
```diff
@@ -3578,9 +3570,8 @@ extern (C++) final class ArrayLiteralExp : Expression
             {
                 return false;
             }
-            for (size_t i = 0; i < elements.dim; i++)
+            foreach (i, e1; *elements)
             {
-                Expression e1 = (*elements)[i];
                 Expression e2 = (*ae.elements)[i];
```
Comment (wilzbach) [inline src/dmd/expression.d]: ... and another potential `lockstep` one.

File: src/dmd/expression.d [hunk 4]
```diff
@@ -3840,9 +3831,8 @@ extern (C++) final class StructLiteralExp : Expression
                 return false;
             if (elements.dim != se.elements.dim)
                 return false;
-            for (size_t i = 0; i < elements.dim; i++)
+            foreach (i, e1; *elements)
             {
-                Expression e1 = (*elements)[i];
                 Expression e2 = (*se.elements)[i];
```
Comment (wilzbach) [inline src/dmd/expression.d]: Yep, `lockstep` is a common pattern.

File: src/dmd/expression.d [hunk 5]
```diff
@@ -4791,8 +4779,8 @@ extern (C++) final class IsExp : Expression
         {
             p = new TemplateParameters();
             p.setDim(parameters.dim);
-            for (size_t i = 0; i < p.dim; i++)
-                (*p)[i] = (*parameters)[i].syntaxCopy();
+            foreach (i, el; *parameters)
```
Comment (wilzbach) [inline src/dmd/expression.d]: `foreach (e, ref e2; *parameters.lockstep(*p))` ...

**Review by jacob-carlborg — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Comment by WalterBright** [comment]
It's not straightforward at all. Many of the loops change the array they are looping over, which doesn't work if it's converted to foreach. That's likely why this PR is failing.

**Pull request merged**